### PR TITLE
72423 attachment as a recorded preservation value

### PR DESCRIPTION
**Pull request by janid1967**
More unit tests will come in later commits today

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/DeleteAttachment/DeleteFieldValueAttachmentCommand.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using MediatR;
+using ServiceResult;
+
+namespace Equinor.Procosys.Preservation.Command.RequirementCommands.DeleteAttachment
+{
+    // unit test accessvalidator
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementCommands/DeleteAttachment/DeleteFieldValueAttachmentCommand.cs:6]: remove comment

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/DeleteAttachment/DeleteFieldValueAttachmentCommandValidator.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                .MustAsync((command, token) => HasRequirementWithActivePeriodAsync(command.TagId, command.RequirementId, token))
+                .WithMessage(command =>
+                    $"Tag doesn't have this requirement with active period! Tag={command.TagId}. Requirement={command.RequirementId}")
+                .MustAsync((command, token) => BeAFieldForAttachmentAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field values can not be recorded for field type! Field={command.FieldId}")
+                .MustAsync((command, token) => BeAnExistingFieldAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field doesn't exists! Field={command.FieldId}")
+                .MustAsync((command, token) => NotBeAVoidedFieldAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field is voided! Field={command.FieldId}");
+
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementCommands/DeleteAttachment/DeleteFieldValueAttachmentCommandValidator.cs:35]: remove extra blank line

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandValidator.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                .MustAsync((command, token) => HasRequirementWithActivePeriodAsync(command.TagId, command.RequirementId, token))
+                .WithMessage(command =>
+                    $"Tag doesn't have this requirement with active period! Tag={command.TagId}. Requirement={command.RequirementId}")
+                .MustAsync((command, token) => BeAFieldForAttachmentAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field values can not be recorded for field type! Field={command.FieldId}")
+                .MustAsync((command, token) => BeAnExistingFieldAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field doesn't exists! Field={command.FieldId}")
+                .MustAsync((command, token) => NotBeAVoidedFieldAsync(command.FieldId, token))
+                .WithMessage((command) => $"Field is voided! Field={command.FieldId}");
+
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandValidator.cs:35]: remove

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        public async Task<Result<int>> Handle(UploadFieldValueAttachmentCommand request, CancellationToken cancellationToken)
+        {
+            var tag = await _projectRepository.GetTagByTagIdAsync(request.TagId);
+            var requirement = tag.Requirements.Single(r => r.Id == request.RequirementId);
+
+            var requirementDefinition =
+                await _requirementTypeRepository.GetRequirementDefinitionByIdAsync(requirement.RequirementDefinitionId);
+
+            var attachment = requirement.GetAlreadyRecordedAttachment(request.FieldId, requirementDefinition);
+
```
Comment (llykkelarsen) [inline src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs:49]: Should rowversion be implemented for attachment? If we both upload an attachment on the same instance, whoever is last will delete the other's?

File: src/Equinor.Procosys.Preservation.Domain/AggregateModels/ProjectAggregate/FieldValue.cs [hunk 5]
```diff
@@ -26,6 +26,8 @@ protected FieldValue(string plant, Field field)
         public DateTime CreatedAtUtc { get; private set; }
         public int CreatedById { get; private set; }
 
+        public FieldValueAttachment FieldValueAttachment { get; set; }
```
[the remark this answers is not in the document]
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Domain/AggregateModels/ProjectAggregate/FieldValue.cs:29]: Nice catch

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs (hunk 4 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs:49]: I've thought about it. We discussed it with Pål last week. We found that is a minimum chance that 2 end users preserve same tags at very same time. You'er right, but we wait implementing it

Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs:49]: Wonf fix

Comment (llykkelarsen) [reply src/Equinor.Procosys.Preservation.Command/RequirementCommands/Upload/UploadFieldValueAttachmentCommandHandler.cs:49]: OK

**Pull request merged**